File: services/ai-agent-py/converact_env.py
"""Fail-closed compatibility boundary for Converact environment variables."""

from __future__ import annotations

import json
import re
import warnings
from collections.abc import Callable, Mapping, MutableMapping

Environment = Mapping[str, str | None]
MutableEnvironment = MutableMapping[str, str | None]
DeprecationEvent = dict[str, str]
DeprecationHandler = Callable[[DeprecationEvent], None]

_ENV_SUFFIX = re.compile(r"^[A-Z][A-Z0-9_]*$")


def _keys(scope: str, suffix: str) -> tuple[str, str]:
    if not _ENV_SUFFIX.fullmatch(suffix):
        raise ValueError(f"invalid branded environment variable suffix: {suffix}")
    if scope == "fabric":
        return f"CONVERACT_FABRIC_{suffix}", f"OPC_IVEKIT_{suffix}"
    return f"CONVERACT_{suffix}", f"OPC_{suffix}"


def _default_deprecation_handler(event: DeprecationEvent) -> None:
    warnings.warn(
        json.dumps(event, separators=(",", ":"), sort_keys=True),
        DeprecationWarning,
        stacklevel=3,
    )


def _resolve(
    env: Environment,
    scope: str,
    suffix: str,
    on_deprecation: DeprecationHandler | None,
) -> str | None:
    current_key, legacy_key = _keys(scope, suffix)
    has_current = current_key in env
    has_legacy = legacy_key in env

    if has_current and has_legacy and env[current_key] != env[legacy_key]:
        raise ValueError(
            "conflicting branded environment variables: "
            f"{current_key} and {legacy_key}"
        )
    if has_current:
        return env[current_key]
    if not has_legacy:
        return None

    (on_deprecation or _default_deprecation_handler)(
        {
            "event": "converact.config.deprecated_environment_key",
            "scope": scope,
            "current_key": current_key,
            "legacy_key": legacy_key,
        }
    )
    return env[legacy_key]


def resolve_brand_env(
    env: Environment,
    suffix: str,
    *,
    on_deprecation: DeprecationHandler | None = None,
) -> str | None:
    return _resolve(env, "brand", suffix, on_deprecation)


def resolve_fabric_env(
    env: Environment,
    suffix: str,
    *,
    on_deprecation: DeprecationHandler | None = None,
) -> str | None:
    return _resolve(env, "fabric", suffix, on_deprecation)


def resolve_converact_env(
    env: Environment,
    key: str,
    *,
    on_deprecation: DeprecationHandler | None = None,
) -> str | None:
    if key.startswith("CONVERACT_FABRIC_"):
        return resolve_fabric_env(
            env,
            key.removeprefix("CONVERACT_FABRIC_"),
            on_deprecation=on_deprecation,
        )
    if key.startswith("OPC_IVEKIT_"):
        return resolve_fabric_env(
            env,
            key.removeprefix("OPC_IVEKIT_"),
            on_deprecation=on_deprecation,
        )
    if key.startswith("CONVERACT_"):
        return resolve_brand_env(
            env,
            key.removeprefix("CONVERACT_"),
            on_deprecation=on_deprecation,
        )
    if key.startswith("OPC_"):
        return resolve_brand_env(
            env,
            key.removeprefix("OPC_"),
            on_deprecation=on_deprecation,
        )
    return env[key] if key in env else None


def install_brand_env_aliases(
    env: MutableEnvironment,
    *,
    on_deprecation: DeprecationHandler | None = None,
) -> dict[str, list[str]]:
    aliases: dict[str, tuple[str, str]] = {}
    for key in env:
        if key.startswith("CONVERACT_FABRIC_"):
            aliases[key] = ("fabric", key.removeprefix("CONVERACT_FABRIC_"))
        elif key.startswith("OPC_IVEKIT_"):
            suffix = key.removeprefix("OPC_IVEKIT_")
            aliases[f"CONVERACT_FABRIC_{suffix}"] = ("fabric", suffix)
        elif key.startswith("CONVERACT_"):
            aliases[key] = ("brand", key.removeprefix("CONVERACT_"))
        elif key.startswith("OPC_"):
            suffix = key.removeprefix("OPC_")
            aliases[f"CONVERACT_{suffix}"] = ("brand", suffix)

    pending: list[tuple[str, str | None]] = []
    deprecations: list[DeprecationEvent] = []
    for current_key in sorted(aliases):
        scope, suffix = aliases[current_key]
        had_current = current_key in env
        value = _resolve(env, scope, suffix, deprecations.append)
        if not had_current:
            pending.append((current_key, value))

    for current_key, value in pending:
        env[current_key] = value
    emit = on_deprecation or _default_deprecation_handler
    for event in deprecations:
        emit(event)

    return {"installed": [current_key for current_key, _ in pending]}

File: src/config/converact-env.ts
export type BrandEnvScope = 'brand' | 'fabric';

export interface BrandEnvDeprecationEvent {
  event: 'converact.config.deprecated_environment_key';
  scope: BrandEnvScope;
  current_key: string;
  legacy_key: string;
}

export interface BrandEnvOptions {
  onDeprecation?: (event: BrandEnvDeprecationEvent) => void;
}

export interface InstalledBrandEnvAliases {
  installed: string[];
}

type Environment = Readonly<Record<string, string | undefined>>;
type MutableEnvironment = Record<string, string | undefined>;

const ENV_SUFFIX = /^[A-Z][A-Z0-9_]*$/;
const emittedLegacyKeys = new Set<string>();

function owns(env: Environment, key: string): boolean {
  return Object.prototype.hasOwnProperty.call(env, key);
}

function defaultDeprecationEmitter(event: BrandEnvDeprecationEvent): void {
  if (emittedLegacyKeys.has(event.legacy_key)) return;
  emittedLegacyKeys.add(event.legacy_key);
  process.emitWarning(JSON.stringify(event), {
    code: 'CONVERACT_DEPRECATED_ENV',
  });
}

function keysFor(scope: BrandEnvScope, suffix: string): {
  currentKey: string;
  legacyKey: string;
} {
  if (!ENV_SUFFIX.test(suffix)) {
    throw new Error(`invalid branded environment variable suffix: ${suffix}`);
  }

  return scope === 'fabric'
    ? {
        currentKey: `CONVERACT_FABRIC_${suffix}`,
        legacyKey: `OPC_IVEKIT_${suffix}`,
      }
    : {
        currentKey: `CONVERACT_${suffix}`,
        legacyKey: `OPC_${suffix}`,
      };
}

function resolve(
  env: Environment,
  scope: BrandEnvScope,
  suffix: string,
  options: BrandEnvOptions,
): string | undefined {
  const { currentKey, legacyKey } = keysFor(scope, suffix);
  const hasCurrent = owns(env, currentKey);
  const hasLegacy = owns(env, legacyKey);

  if (hasCurrent && hasLegacy && env[currentKey] !== env[legacyKey]) {
    throw new Error(
      `conflicting branded environment variables: ${currentKey} and ${legacyKey}`,
    );
  }

  if (hasCurrent) return env[currentKey];
  if (!hasLegacy) return undefined;

  (options.onDeprecation ?? defaultDeprecationEmitter)({
    event: 'converact.config.deprecated_environment_key',
    scope,
    current_key: currentKey,
    legacy_key: legacyKey,
  });
  return env[legacyKey];
}

export function resolveBrandEnv(
  env: Environment,
  suffix: string,
  options: BrandEnvOptions = {},
): string | undefined {
  return resolve(env, 'brand', suffix, options);
}

export function resolveFabricEnv(
  env: Environment,
  suffix: string,
  options: BrandEnvOptions = {},
): string | undefined {
  return resolve(env, 'fabric', suffix, options);
}

export function resolveConveractEnv(
  env: Environment,
  key: string,
  options: BrandEnvOptions = {},
): string | undefined {
  if (key.startsWith('CONVERACT_FABRIC_')) {
    return resolveFabricEnv(env, key.slice('CONVERACT_FABRIC_'.length), options);
  }
  if (key.startsWith('OPC_IVEKIT_')) {
    return resolveFabricEnv(env, key.slice('OPC_IVEKIT_'.length), options);
  }
  if (key.startsWith('CONVERACT_')) {
    return resolveBrandEnv(env, key.slice('CONVERACT_'.length), options);
  }
  if (key.startsWith('OPC_')) {
    return resolveBrandEnv(env, key.slice('OPC_'.length), options);
  }
  return owns(env, key) ? env[key] : undefined;
}

export function installBrandEnvAliases(
  env: MutableEnvironment = process.env,
  options: BrandEnvOptions = {},
): InstalledBrandEnvAliases {
  const aliases = new Map<string, { scope: BrandEnvScope; suffix: string }>();

  for (const key of Object.keys(env)) {
    if (key.startsWith('CONVERACT_FABRIC_')) {
      aliases.set(key, { scope: 'fabric', suffix: key.slice('CONVERACT_FABRIC_'.length) });
    } else if (key.startsWith('OPC_IVEKIT_')) {
      const suffix = key.slice('OPC_IVEKIT_'.length);
      aliases.set(`CONVERACT_FABRIC_${suffix}`, { scope: 'fabric', suffix });
    } else if (key.startsWith('CONVERACT_')) {
      aliases.set(key, { scope: 'brand', suffix: key.slice('CONVERACT_'.length) });
    } else if (key.startsWith('OPC_')) {
      const suffix = key.slice('OPC_'.length);
      aliases.set(`CONVERACT_${suffix}`, { scope: 'brand', suffix });
    }
  }

  const pending: Array<{ currentKey: string; value: string | undefined }> = [];
  const deprecations: BrandEnvDeprecationEvent[] = [];

  for (const [currentKey, alias] of [...aliases].sort(([left], [right]) =>
    left.localeCompare(right),
  )) {
    const hadCurrent = owns(env, currentKey);
    const value = resolve(env, alias.scope, alias.suffix, {
      onDeprecation: (event) => deprecations.push(event),
    });
    if (!hadCurrent) pending.push({ currentKey, value });
  }

  for (const { currentKey, value } of pending) env[currentKey] = value;
  const emit = options.onDeprecation ?? defaultDeprecationEmitter;
  for (const event of deprecations) emit(event);

  return { installed: pending.map(({ currentKey }) => currentKey) };
}
